File: ts/streaming-kbn/raw-moments-klein-kbn.ts
import { KleinKbnAccumulator } from './klein-kbn-accumulator';

/**
 * Streaming mean, variance, skewness, and kurtosis via raw power sums (x¹..x⁴)
 * with KBN double-compensated accumulation.
 *
 * Accumulates Σx, Σx², Σx³, Σx⁴ using KleinKbnAccumulator for each,
 * plus a separate Welford-style variance tracker (also KBN-compensated).
 * Raw sums are converted to central moments at query time.
 *
 * Supports both LIFO revert and FIFO rolling window (via revert/update cycle).
 */
export class RawMomentsKleinKbn {
    private _n = 0;
    private readonly _x1 = new KleinKbnAccumulator();
    private readonly _x2 = new KleinKbnAccumulator();
    private readonly _x3 = new KleinKbnAccumulator();
    private readonly _x4 = new KleinKbnAccumulator();
    private readonly _mean = new KleinKbnAccumulator();
    private readonly _s = new KleinKbnAccumulator();
    private readonly _ddof: number;
    private readonly _bias: boolean;
    private readonly _fisher: boolean;

    constructor(ddof = 1, bias = true, fisher = true) {
        this._ddof = ddof;
        this._bias = bias;
        this._fisher = fisher;
    }

    /** Clears all accumulated state. */
    reset(): void {
        this._n = 0;
        this._x1.reset();
        this._x2.reset();
        this._x3.reset();
        this._x4.reset();
        this._mean.reset();
        this._s.reset();
    }

    /** Adds a new sample x to the accumulator. */
    update(x: number): void {
        this._n++;
        this._x1.update(x);
        const x2 = x * x;
        this._x2.update(x2);
        const x3 = x2 * x;
        this._x3.update(x3);
        const x4 = x3 * x;
        this._x4.update(x4);

        const n = this._n;
        const delta = x - this._mean.value;
        this._mean.update(delta / n);
        this._s.update(delta * (x - this._mean.value));
    }

    /** Removes a previously added sample x from the accumulator. */
    revert(x: number): void {
        this._n--;
        this._x1.revert(x);
        const x2 = x * x;
        this._x2.revert(x2);
        const x3 = x2 * x;
        this._x3.revert(x3);
        const x4 = x3 * x;
        this._x4.revert(x4);

        const delta = x - this._mean.value;
        const n = this._n;
        this._mean.revert(delta / n);
        this._s.revert(delta * (x - this._mean.value));
    }

    /** Returns the current arithmetic mean. */
    get mean(): number {
        return this._mean.value;
    }

    /** Returns the current sample count. */
    get n(): number {
        return this._n;
    }

    /** Returns the current variance. Returns NaN if n <= ddof. */
    get variance(): number {
        const n = this._n - this._ddof;
        if (n <= 0) return NaN;
        const sv = this._s.value;
        if (sv < 0) {
            this._s.reset();
            return NaN;
        }
        return sv / n;
    }

    /** Returns the current standard deviation. Returns NaN if n <= ddof. */
    get standardDeviation(): number {
        const n = this._n - this._ddof;
        if (n <= 0) return NaN;
        return Math.sqrt(this._s.value / n);
    }

    /** Returns the current skewness. Returns NaN if n < 3. */
    get skewness(): number {
        const N = this._n;
        if (N < 3) return NaN;
        const a = this._x1.value / N;
        const b = this._x2.value / N - a * a;
        if (b <= 1e-14) return NaN;
        const r = Math.sqrt(b);
        const c = this._x3.value / N - a * a * a - 3 * a * b;
        const g1 = c / (r * r * r);
        if (this._bias) return g1;
        return g1 * Math.sqrt(N * (N - 1)) / (N - 2);
    }

    /** Returns the current kurtosis. Returns NaN if n < 4. */
    get kurtosis(): number {
        const N = this._n;
        if (N < 4) return NaN;
        const a = this._x1.value / N;
        let r = a * a;
        const b = this._x2.value / N - r;
        if (b <= 1e-14) return NaN;
        r *= a;
        const c = this._x3.value / N - r - 3 * a * b;
        r *= a;
        const d = this._x4.value / N - r - 6 * b * a * a - 4 * c * a;
        const raw = d / (b * b);
        if (!this._bias) {
            const adj = ((N * N - 1) * raw - 3 * (N - 1) * (N - 1)) / ((N - 2) * (N - 3));
            return this._fisher ? adj : adj + 3.0;
        }
        return this._fisher ? raw - 3.0 : raw;
    }
}

File: go/streamingkbn/rawmomentskleinkbn.go
package streamingkbn

import "math"

// RawMomentsKleinKBN provides streaming mean, variance, skewness, and kurtosis
// via raw power sums (x¹..x⁴) with KBN (Kahan-Babuška-Neumaier) double-compensated
// accumulation.
//
// It accumulates Σx, Σx², Σx³, Σx⁴ using KleinKBNAccumulator for each,
// plus a separate Welford-style variance tracker (also KBN-compensated).
// Raw sums are converted to central moments at query time.
//
// Supports both LIFO revert (undo the most recent update) and FIFO
// rolling window (via the revert/update cycle) because subtracting
// from a linear sum preserves the KBN compensation state.
//
// Parameters:
//   - ddof: Delta degrees of freedom for variance.
//     variance = m2 / (n - ddof). ddof=0 gives population, ddof=1 gives sample.
//   - bias: If true, compute population standardized moments (m3/m2^1.5).
//     If false, apply the Fisher-Pearson adjusted (bias-corrected) factor:
//     skewness_bcf = skewness_pop * sqrt(n*(n-1)) / (n-2)
//   - fisher: If true, return excess kurtosis (subtract 3 so Gaussian→0).
//     If false, return raw kurtosis (Gaussian→3).
//     Applied after the bias correction when bias=false.
//
// Formulas:
//
// Skewness (bias=true):
//
//	g1 = sqrt(n) * m3 / m2^1.5
//
// Skewness (bias=false):
//
//	G1 = g1 * sqrt(n*(n-1)) / (n-2)
//
// Kurtosis (bias=true, fisher=true):
//
//	g2 = n * m4 / m2^2 - 3
//
// Kurtosis (bias=true, fisher=false):
//
//	g2 = n * m4 / m2^2
//
// Kurtosis (bias=false, fisher=true):
//
//	G2 = ((n^2-1) * n*m4/m2^2 - 3*(n-1)^2) / ((n-2)*(n-3))
//
// Kurtosis (bias=false, fisher=false):
//
//	G2 = ((n^2-1) * n*m4/m2^2 - 3*(n-1)^2) / ((n-2)*(n-3)) + 3
type RawMomentsKleinKBN struct {
	n      int
	x1     *KleinKBNAccumulator
	x2     *KleinKBNAccumulator
	x3     *KleinKBNAccumulator
	x4     *KleinKBNAccumulator
	mean   *KleinKBNAccumulator
	s      *KleinKBNAccumulator
	ddof   int
	bias   bool
	fisher bool
}

// NewRawMomentsKleinKBN creates a new RawMomentsKleinKBN with the given parameters.
func NewRawMomentsKleinKBN(ddof int, bias, fisher bool) *RawMomentsKleinKBN {
	return &RawMomentsKleinKBN{
		x1:     &KleinKBNAccumulator{},
		x2:     &KleinKBNAccumulator{},
		x3:     &KleinKBNAccumulator{},
		x4:     &KleinKBNAccumulator{},
		mean:   &KleinKBNAccumulator{},
		s:      &KleinKBNAccumulator{},
		ddof:   ddof,
		bias:   bias,
		fisher: fisher,
	}
}

// Reset clears all accumulated state.
func (m *RawMomentsKleinKBN) Reset() {
	m.n = 0
	m.x1.Reset()
	m.x2.Reset()
	m.x3.Reset()
	m.x4.Reset()
	m.mean.Reset()
	m.s.Reset()
}

// Update adds a new sample x to the accumulator.
func (m *RawMomentsKleinKBN) Update(x float64) {
	m.n++
	m.x1.Update(x)
	x2 := x * x
	m.x2.Update(x2)
	x3 := x2 * x
	m.x3.Update(x3)
	x4 := x3 * x
	m.x4.Update(x4)

	n := float64(m.n)
	delta := x - m.mean.Value()
	m.mean.Update(delta / n)
	m.s.Update(delta * (x - m.mean.Value()))
}

// Revert removes a previously added sample x from the accumulator.
func (m *RawMomentsKleinKBN) Revert(x float64) {
	m.n--
	m.x1.Revert(x)
	x2 := x * x
	m.x2.Revert(x2)
	x3 := x2 * x
	m.x3.Revert(x3)
	x4 := x3 * x
	m.x4.Revert(x4)

	delta := x - m.mean.Value()
	n := float64(m.n)
	m.mean.Revert(delta / n)
	m.s.Revert(delta * (x - m.mean.Value()))
}

// Mean returns the current arithmetic mean.
func (m *RawMomentsKleinKBN) Mean() float64 {
	return m.mean.Value()
}

// Variance returns the current variance.
// Returns NaN if n <= ddof.
func (m *RawMomentsKleinKBN) Variance() float64 {
	n := float64(m.n - m.ddof)
	if n <= 0 {
		return math.NaN()
	}
	s := m.s.Value()
	if s < 0 {
		m.s.Reset()
		return math.NaN()
	}
	return s / n
}

// StandardDeviation returns the current standard deviation.
// Returns NaN if n <= ddof.
func (m *RawMomentsKleinKBN) StandardDeviation() float64 {
	n := float64(m.n - m.ddof)
	if n <= 0 {
		return math.NaN()
	}
	return math.Sqrt(m.s.Value() / n)
}

// Skewness returns the current skewness.
// Returns NaN if n < 3.
func (m *RawMomentsKleinKBN) Skewness() float64 {
	N := float64(m.n)
	if m.n < 3 {
		return math.NaN()
	}
	A := m.x1.Value() / N
	B := m.x2.Value()/N - A*A
	if B <= 1e-14 {
		return math.NaN()
	}
	R := math.Sqrt(B)
	C := m.x3.Value()/N - A*A*A - 3*A*B
	g1 := C / (R * R * R)
	if m.bias {
		return g1
	}
	return g1 * math.Sqrt(N*(N-1)) / (N - 2)
}

// Kurtosis returns the current kurtosis.
// Returns NaN if n < 4.
func (m *RawMomentsKleinKBN) Kurtosis() float64 {
	N := float64(m.n)
	if m.n < 4 {
		return math.NaN()
	}
	A := m.x1.Value() / N
	R := A * A
	B := m.x2.Value()/N - R
	if B <= 1e-14 {
		return math.NaN()
	}
	R *= A
	C := m.x3.Value()/N - R - 3*A*B
	R *= A
	D := m.x4.Value()/N - R - 6*B*A*A - 4*C*A
	raw := D / (B * B)
	if !m.bias {
		adj := ((N*N-1)*raw - 3*(N-1)*(N-1)) / ((N - 2) * (N - 3))
		if m.fisher {
			return adj
		}
		return adj + 3.0
	}
	if m.fisher {
		return raw - 3.0
	}
	return raw
}
